### ISSUE TITLE
EF Core, CI/CD, and Migrations w/ breaking changes

### ISSUE THREAD

**Issue by cmullendore** [dotnet/efcore#28494]
tl;dr: What is the best strategy for managing an EF core project in a CI/CD world when schema changes are breaking?

For example, you have a deployed model, code-first, and you're using migrations. You have a new product version you want to test and deploy, but it includes a schema migration that is a breaking change. For example, converting an integer column or GUID column to a string column. Or perhaps adding a new column as required (which translates to the DB) that the old code/schema isn't aware of. 

In these instances, the "old" code won't work with the new change, and the "new" code can't work with the existing (pre-migration) schema. The old code doesn't know about the now mandatory, non-null column, and the new code can't work without the field present in the DB/schema.

Doing a rollout means ensuring that the previous iteration of the product continues to run without failure while deploying and warming up the next version of the product. If I do a migration, the previous version breaks... If I don't do the migration, the new version breaks.

What is the team's best recommendation for working with this?

(first post, please feel free to point me at an existing answer. Thanks)

**Comment by ErikEJ** [comment]
Use development patterns and scripts that avoids breaking changes - for example, add new columns with default values or add data migration scripts, and deploy the new column (and an app that ignores the old column) all the way to production - then do another deployment that removes the old column from the database.

**Comment by ajcvickers** [comment]
@cmullendore To expand on what Erik said, there is no way to do this without explicitly planning for it in advance and using a multi-step process which does not break running apps. However, I would also really consider if you need to have the app continue running during an upgrade. For some cases, like a service exposed to external consumers, there is no way to stop access to the service, then something like this must be done. But given that doing this generally costs a lot in terms of development time, if it is possible to have a maintenance window where the app is temporarily unavailable, then this is generally _much_ cheaper and simpler.

**Comment by cmullendore** [comment]
Thank you both. I guess in my brain I hadn't made the jump to a multi-deployment process to handle this even remotely gracefully. @ajcvickers respectfully, while I get that downtime makes pretty much everything easier from a deployment process and while this might be achievable in perhaps an internal corporate setting, I don't think it's realistic in an internet-facing process. Publicly, downtime, scheduled or otherwise, is pretty much all-cost avoided.

So this is in essence what's being suggested:

1. Plan, plan, plan for what v2 requires.
2. Do a perhaps schema-migration-only deployment of v1, incorporating necessary schema changes for v2. In this step we might skip certain things like referential integrity on the columns being created since they won't necessarily have data.
3. Optionally, do a v1 that does include any necessary data migration/transforms. In this instance v1 may not offer the new functionality, but might at least implement persistence processes that align with v2's needs such that when v2 is deployed all of the data is ready.
4. Prepare and deploy v2 in such a way as it doesn't necessarily light up the new v2 features yet but is still in essence v1 but ready to be feature-flipped. Because at this point all of the data structures are ready to go, v2 should flip on seamlessly with no downtime. Ideally at this point v2 should require _zero_ schema mods. 
5. Do whatever feature-flipping might be necessary to activate v2 features. Because the data has been aligned since v1, enabling v2 features should be a no-brainer. For revertability and/or co-existence with v1's features still not being 100% lit, the new v2 must also ensure data structures compatible with v1 are maintained alongside v2's data persistence.
6. Complete feature flip and stabilize to v2. 
7. Iterate v2.1 to refactor any compatibility code with v1 out of the model, deploy, and initiate a migration to purge legacy data, since at this point we're at a fix-forward situation anyway.

Would love to hear if this is overly-planned or if I'm missing anything.

**Comment by ajcvickers** [comment]
@cmullendore I'm not an expert in this area, and it's not really a EF or even .NET-specific thing. A quick search brought up this [Update your Database Schema Without Downtime](https://thorben-janssen.com/update-database-schema-without-downtime/), which seems to be a pretty good overview.
